### ISSUE TITLE
Redis Cache TTL problem

### ISSUE THREAD

**Issue by cat24max** [santigarcor/laratrust#280]
- Laravel Version: 5.5
- Laratrust Version: 3.0

Hey guys,
before I start: Yes, I know 3.0 is a very old version, but I didn't manage to update until now.

My application is getting hit very hard compared to a few months back and I used the "array" driver for cache until now.
Yesterday I switched this to Redis and it worked fine for sessions and everything else I cache.

BUT when it comes to Laratrust, the TTL seems to somehow destroy permissions. My users told me that everything worked normal for a time, but after an hour or so (60 minutes TTL) groups where now able to access things in the application they were not allowed to. By saving the group again (and hence resetting the cache) everything was fine again until the TTL runs out.

Any idea how that happens or how I could fix it?

**Comment by santigarcor** [comment]
Try overriding in your user model the `cachedRoles` method so you don't use cache for them. Also, in your  role model override the `cachedPermissions` method.

**Comment by cat24max** [comment]
Yes I guess that would work, but wouldn't this increase load on my database server? The whole point of me enabling Redis is to decrease the DB load :)

**Comment by cat24max** [comment]
Ok, this issue can be closed and is kind of my fault, though maybe it's still helpful for someone else.

When I developed my app, it was standalone for 1 customer. But in March this year I converted it to a multi-tenant apps, using separated groups and permissions for each customer. But the cache key didn't contain the tenant ID, so they would overwrite themselves.
This wasn't an issue while using the array driver, but was after changing to Redis.

So the simple fix was just making the cache key unique by tenant (by overwriting the methods from the User and Role Trait)
`$cacheKey = 'laratrust_'.config('app.tenant_id').'_permissions_for_role_' . $this->getKey();`
